blockindex/liquidstaking_indexer.go

```diff
@@ -0,0 +1,967 @@
+	ErrBucketInfoNotExist = errors.New("bucket info does not exist")
+)
+
+func init() {
+	var err error
+	_liquidStakingInterface, err = abi.JSON(strings.NewReader(LiquidStakingContractABI))
+	if err != nil {
+		panic(err)
+	}
+}
+
+// NewLiquidStakingIndexer creates a new liquid staking indexer
+func NewLiquidStakingIndexer(kvStore db.KVStore, blockInterval time.Duration, candNameToOwnerFunc candNameToOwnerFunc) LiquidStakingIndexer {
+	return &liquidStakingIndexer{
+		blockInterval:       blockInterval,
+		kvstore:             kvStore,
+		cache:               newLiquidStakingCache(),
+		candNameToOwnerFunc: candNameToOwnerFunc,
+	}
+}
+
+// Start starts the indexer
+func (s *liquidStakingIndexer) Start(ctx context.Context) error {
+	if err := s.kvstore.Start(ctx); err != nil {
+		return err
+	}
+	return s.loadCache()
+}
+
+// Stop stops the indexer
+func (s *liquidStakingIndexer) Stop(ctx context.Context) error {
+	if err := s.kvstore.Stop(ctx); err != nil {
+		return err
+	}
+	return nil
+}
+
+// PutBlock puts a block into indexer
+func (s *liquidStakingIndexer) PutBlock(ctx context.Context, blk *block.Block) error {
+	// new dirty cache
+	dirty := newLiquidStakingDirty(s.cache)
+	dirty.putHeight(blk.Height())
+	// make action map
+	actionMap := make(map[hash.Hash256]*action.SealedEnvelope)
+	for i := range blk.Actions {
+		h, err := blk.Actions[i].Hash()
+		if err != nil {
+			return err
+		}
+		actionMap[h] = &blk.Actions[i]
+	}
+
+	// handle events of block
+	for _, receipt := range blk.Receipts {
+		if receipt.Status != uint64(iotextypes.ReceiptStatus_Success) {
+			continue
+		}
+		act, ok := actionMap[receipt.ActionHash]
+		if !ok {
+			return errors.Errorf("action %x not found", receipt.ActionHash)
+		}
+		for _, log := range receipt.Logs() {
+			if log.Address != LiquidStakingContractAddress {
+				continue
+			}
+			if err := s.handleEvent(ctx, dirty, blk, act, log); err != nil {
+				return err
+			}
+		}
+	}
+
+	// commit dirty cache
+	return s.commit(dirty)
+}
+
+// DeleteTipBlock deletes the tip block from indexer
+func (s *liquidStakingIndexer) DeleteTipBlock(context.Context, *block.Block) error {
+	return errors.New("not implemented")
+}
+
+// Height returns the tip block height
+func (s *liquidStakingIndexer) Height() (uint64, error) {
+	return s.cache.getHeight(), nil
+}
+
+// CandidateVotes returns the candidate votes
+func (s *liquidStakingIndexer) CandidateVotes(ownerAddr string) *big.Int {
+	return s.cache.getCandidateVotes(ownerAddr)
+}
+
+// Buckets returns the buckets
+func (s *liquidStakingIndexer) Buckets() ([]*Bucket, error) {
+	vbs := []*Bucket{}
+	for id, bi := range s.cache.idBucketMap {
+		bt := s.cache.mustGetBucketType(bi.TypeIndex)
+		vb, err := s.convertToVoteBucket(id, bi, bt)
+		if err != nil {
+			return nil, err
+		}
+		vbs = append(vbs, vb)
+	}
+	return vbs, nil
+}
+
+// Bucket returns the bucket
+func (s *liquidStakingIndexer) Bucket(id uint64) (*Bucket, error) {
```

Comment:
The implementation of the indexer needs to consider multi-threads scenario. Currently, the cache will panic when Bucket() and PutBlock() are called at the same time.

Reply:
Indeed, a great suggestion. I have recorded it as a to-do item, and it will be addressed in the next PR.

Reply:
already fixed at the latest commit